sdks/java/io/neo4j/src/main/java/org/apache/beam/sdk/io/neo4j/Neo4jIO.java

```diff
@@ -0,0 +1,1346 @@
+ *    );
+ * }</pre>
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class Neo4jIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Neo4jIO.class);
+
+  /**
+   * Read all rows using a Neo4j Cypher query.
+   *
+   * @param <ParameterT> Type of the data representing query parameters.
+   * @param <OutputT> Type of the data to be read.
+   */
+  public static <ParameterT, OutputT> ReadAll<ParameterT, OutputT> readAll() {
+    return new AutoValue_Neo4jIO_ReadAll.Builder<ParameterT, OutputT>()
+        .setFetchSize(ValueProvider.StaticValueProvider.of(Config.defaultConfig().fetchSize()))
+        .build();
+  }
+
+  /**
+   * Write all rows using a Neo4j Cypher UNWIND cypher statement. This sets a default batch batch
+   * size of 5000.
+   *
+   * @param <ParameterT> Type of the data representing query parameters.
+   */
+  public static <ParameterT> WriteUnwind<ParameterT> writeUnwind() {
+    return new AutoValue_Neo4jIO_WriteUnwind.Builder<ParameterT>()
+        .setBatchSize(ValueProvider.StaticValueProvider.of(5000L))
+        .build();
+  }
+
+  private static <ParameterT, OutputT> PCollection<OutputT> getOutputPCollection(
+      PCollection<ParameterT> input, DoFn<ParameterT, OutputT> writeFn, Coder<OutputT> coder) {
+    PCollection<OutputT> output = input.apply(ParDo.of(writeFn)).setCoder(coder);
+
+    try {
+      TypeDescriptor<OutputT> typeDesc = coder.getEncodedTypeDescriptor();
+      SchemaRegistry registry = input.getPipeline().getSchemaRegistry();
+      Schema schema = registry.getSchema(typeDesc);
+      output.setSchema(
+          schema,
+          typeDesc,
+          registry.getToRowFunction(typeDesc),
+          registry.getFromRowFunction(typeDesc));
+    } catch (NoSuchSchemaException e) {
+      // ignore
+    }
+    return output;
+  }
+
+  /**
+   * An interface used by {@link ReadAll} for converting each row of a Neo4j {@link Result} record
+   * {@link Record} into an element of the resulting {@link PCollection}.
+   */
+  @FunctionalInterface
+  public interface RowMapper<T> extends Serializable {
+    T mapRow(Record record) throws Exception;
+  }
+
+  /** This describes all the information needed to create a Neo4j {@link Session}. */
+  @AutoValue
+  public abstract static class DriverConfiguration implements Serializable {
+    public static DriverConfiguration create() {
+      return new AutoValue_Neo4jIO_DriverConfiguration.Builder().build();
+    }
+
+    public static DriverConfiguration create(String url, String username, String password) {
+      checkArgument(url != null, "url can not be null");
+      checkArgument(username != null, "username can not be null");
+      checkArgument(password != null, "password can not be null");
+      return new AutoValue_Neo4jIO_DriverConfiguration.Builder()
+          .build()
+          .withUrl(url)
+          .withUsername(username)
+          .withPassword(password);
+    }
+
+    abstract @Nullable ValueProvider<String> getUrl();
```

Comment:
Thanks for all the feedback and comments.  Unfortunately in this specific case the Neo4j Java driver makes a distinction between encrypted connections, non-encrypted connections and encrypted not specified (e.g. null in our case).  In that last case the driver figures things out automatically.

Reply:
You resolved all the checker framework issues. But there are still several places where a field is marked as nullable even though null values are immediately rejected in the setter method. In these cases we would be better off using AutoValue's generated checks, and remove the nullable annotations since they're misleading.

Reply:
I think it's time for me to admit defeat.  I'm not going to solve this puzzle.  Removing the annotations would in all honesty just re-introduce checker framework errors.  I guess there's a reason why the other IOs simply suppress the warning.

Reply:
Checker framework isn't a heuristic warning, but a modernized type system. Nullability is part of a type or not. You have to decide whether or not it is a valid value at a particular point in the code.

Reply:
Checker (correctly) defaults to assuming any Java method can have arbitrary non-pure behavior. It doesn't have any ability to infer this. And most likely impurity would be contagious enough that such inference would be useless. If you want to indicate that a function will always return the same value you can use [`@Pure` annotation](https://checkerframework.org/api/org/checkerframework/dataflow/qual/Pure.html) which is just a shorthand for `@SideEffectFree` and `@Deterministic`.

That said, this case is very easy. You can just set the return value of `getEncryption()` to a local variable first, so it is known that it cannot change.

Reply:
Also some history: the other IOs suppress nullness checking because I automatically added it to every class that had nullness errors when we made checker on by default. New code should never suppress it.

Reply:
I'll let Sachin know I'm done.  There's no way I'll be able to explain all this to whoever I'm handing over the code, let alone maintain it in the future.

Reply:
Yea hopefully for the most part the explanation won't be necessary. Once it is checked in I'd be happy to make a pass to revise stuff so that types match the intention. For example it sounds like encryption is, in fact, optional while perhaps some other fields are not.

Reply:
@ibzib You were right, fields like username and password are allowed to be null but an error was thrown in case you tried to set those to null.  In other case what we're really saying is: the builder can start with a null value since we can't provide a default but you still need to provide a value, non-null if you please.

@kennknowles after struggling with KafkaIO and similar unnecessary checks there I'm now convinced it's really a bad idea to make too many validations.  The use cases for the underlying code vary enough that it becomes hard to predict the special cases.  In the KafkaIO case it is not needed to make explicit usage of the serializers mandatory and in this case, with Neo4jIO it's a mistake to make the Encryption mandatory as it would prevent auto-configuration in the Neo4j Java driver.   As for the NonNull annotations at the level of the builders: they force you to essentially create a method which accepts all mandatory fields totally negating the idea behind the builder in the first place.  If you accept that you're filling in the information one "with" method at a time, you're going to temporary have fields which are null until all information is provided.  In other words, it becomes impossible to use a convenient method like Neo4jIO.readAll() or Neo4jIO.writeUnwind().

In general it's probably most clean if we do the argument checks for the things that really matter in the expand() phase.

Reply:
It's best if we can ensure the configuration never ends up in a bad state (i.e. where required values are set to null).

> As for the NonNull annotations at the level of the builders: they force you to essentially create a method which accepts all mandatory fields totally negating the idea behind the builder in the first place. If you accept that you're filling in the information one "with" method at a time, you're going to temporary have fields which are null until all information is provided.

Your analysis is correct, but I think we should reconsider the assumption that we must only provide `with` methods as the public API. As far as I can tell, there are three things `with` methods accomplish here:

1. null checks
2. helper methods for StaticValueProviders
3. create a modified copy of a configuration

Item 1 is best left to the AutoValue generated code, and 2 is unneeded if we get remove ValueProviders. 3 is the only one that really provides value IMO.

If we make the Builder public, we can remove all the `@Nullable` annotations and null checks. We can still offer `with` methods to address item 3 if we want, but the builder would be the only way to construct an object initially.